### PR TITLE
Ready to Merge: Union tests

### PR DESCRIPTION
**Pull request by rohit-rk**

### CODE REVIEW

File: jx_sqlite/aggs_table.py [hunk 1]
```diff
@@ -360,12 +360,21 @@ def _get(row):
...
                     for json_type, sql in details.sql.items():
-                        concat_sql.append("GROUP_CONCAT(QUOTE(DISTINCT(" + sql + ")))")
+                        concat_sql.append("GROUP_CONCAT(DISTINCT(" + sql + "))")
 
                     if len(concat_sql) > 1:
                         concat_sql = "CONCAT(" + ",".join(concat_sql) + ") AS " + _make_column_name(column_number)
                     else:
                         concat_sql = concat_sql[0] + " AS " + _make_column_name(column_number)
+                    
+                    query_tables = [t["nest"] for t in tables]
```
Comment (klahnakoski) [inline jx_sqlite/aggs_table.py]: I do believe `tables` is `wrap`ed, so you can `query_tables = tables.nest` instead

File: jx_sqlite/aggs_table.py [hunk 2]
```diff
@@ -360,12 +360,21 @@ def _get(row):
                     column_number = len(outer_selects)
 
                     for json_type, sql in details.sql.items():
-                        concat_sql.append("GROUP_CONCAT(QUOTE(DISTINCT(" + sql + ")))")
+                        concat_sql.append("GROUP_CONCAT(DISTINCT(" + sql + "))")
```
Comment (klahnakoski) [inline jx_sqlite/aggs_table.py]: Some quoting will be required in the event there are special characters in the columns

File: jx_sqlite/aggs_table.py [hunk 3]
```diff
@@ -360,12 +360,21 @@ def _get(row):
...
                         concat_sql = concat_sql[0] + " AS " + _make_column_name(column_number)
+                    
+                    query_tables = [t["nest"] for t in tables]
+                    if details.nested_path not in query_tables:
+                        p = details.nested_path
+                        alias = nest_to_alias[p]
+                        from_sql += (
+                            "\nLEFT JOIN\n" + quote_table(concat_field(base_table, p)) + " " + alias +
+                            " ON " + alias + "." + PARENT + " = " + previous.alias + "." + UID
+                        )
```
Comment (klahnakoski) [inline jx_sqlite/aggs_table.py]: This left join might work, but it needs more isolation from the rest of the query:  This left join will multiply with any other left join to create an exponential explosion in the number of records.  Be sure to left join with a table/view/resultset that has only **one row for each UID**.  This means joining with something like:

    (SELECT parent_id, group_concat(var) as unioned FROM nested_table GROUP BY parent_id)

Comment (klahnakoski) [inline jx_sqlite/aggs_table.py]: I added a test to show the potential problem:

https://github.com/mozilla/jx-sqlite/commit/4c24626f7b500a1a53b71cd91f550d3670a85a34

File: jx_sqlite/aggs_table.py (hunk 2 above)
Comment (rohit-rk) [inline jx_sqlite/aggs_table.py]: `NULL`'s are also quoted with this.

File: tests/test_jx/test_edge_1.py [hunk 4]
```diff
@@ -436,15 +436,15 @@ def test_union_values(self):
                 "data": [
                     {"a": "x", "c": [2, 3]},
                     {"a": "y", "c": [4, 5, 6]},
-                    {"a": NULL, "c": [7, 8]}
+                    {"a": NULL, "c": [8, 7]}
```
Comment (rohit-rk) [inline tests/test_jx/test_edge_1.py]: Made these changes in tests to ask if the order should matter in UNION of values.

Comment (klahnakoski) [inline tests/test_jx/test_edge_1.py]: order should not matter with union: union is a set-wise (unordered) operator.

File: jx_sqlite/aggs_table.py (hunk 2 above)
Comment (klahnakoski) [inline jx_sqlite/aggs_table.py]: How about the `json_quote()` function?

Comment (klahnakoski) [inline jx_sqlite/aggs_table.py]: ...plus , if you surround `concat_sql` with `[]`, you can generate JSON, which will be digested nicely with JSON tools on the Python side.

File: tests/test_jx/test_edge_1.py (hunk 4 above)
Comment (klahnakoski) [inline tests/test_jx/test_edge_1.py]: I believe you can use a Python `set` to indicate you do not care about order:

    {"a": NULL, "c": set([8, 7])}

or

    {"a": NULL, "c": {8, 7}}

File: jx_sqlite/aggs_table.py (hunk 2 above)
Comment (klahnakoski) [inline jx_sqlite/aggs_table.py]: ... even better!  `json_group_array()` does all the work for us.

Comment (rohit-rk) [inline jx_sqlite/aggs_table.py]: NULL is still quoted with this.

Comment (rohit-rk) [inline jx_sqlite/aggs_table.py]: As a result of json_group_array(), list in result.data is of type unicode. Not sure how to convert it to list in sql.

```
>>> result.data[0][0]
u'["a","b","c","d","e","f","g","h","i","j","x","y","z"]'


>>> type(result.data[0][0])
<type 'unicode'>Inline images 1
```

Comment (rohit-rk) [inline jx_sqlite/aggs_table.py]: `NULL`'s are still included. I think using where clause to not including them may be a bad logic.

Comment (klahnakoski) [inline jx_sqlite/aggs_table.py]: I believe you change the `pull` function from 

    pull=sql_text_array_to_set(column_number),

to 

    pull=lambda row: json2value(row[column_number])

or something like that

Comment (klahnakoski) [inline jx_sqlite/aggs_table.py]: hopefully the `json2value` will also solve your `NULL` problem

Comment (rohit-rk) [inline jx_sqlite/aggs_table.py]: `NULL` problem is still there

Comment (klahnakoski) [inline jx_sqlite/aggs_table.py]: looking now...

Comment (klahnakoski) [inline jx_sqlite/aggs_table.py]: if you use the `json2value` suggestion I made, the union test passes.  Here is the diff I applied to your branch to make it work.

```diff
 jx_sqlite/aggs_table.py      | 3 ++-
 tests/test_jx/test_edge_1.py | 2 +-
 2 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/jx_sqlite/aggs_table.py b/jx_sqlite/aggs_table.py
index b45f6b8..16c599b 100644
--- a/jx_sqlite/aggs_table.py
+++ b/jx_sqlite/aggs_table.py
@@ -14,6 +14,7 @@ from __future__ import division
 from __future__ import unicode_literals
 
 from mo_dots import listwrap, coalesce, split_field, join_field, startswith_field, relative_field, concat_field
+from mo_json import json2value
 from mo_logs import Log
 from mo_math import Math
 
@@ -399,7 +400,7 @@ class AggsTable(SetOpTable):
                         push_column_name=s.name,
                         push_column=si,
                         push_child=".",
-                        pull=sql_text_array_to_set(column_number),
+                        pull=lambda row: json2value(row[column_number]),
                         sql=concat_sql,
                         column_alias=_make_column_name(column_number),
                         type=sql_type_to_json_type[json_type]
diff --git a/tests/test_jx/test_edge_1.py b/tests/test_jx/test_edge_1.py
index 16ee31a..3b12512 100644
--- a/tests/test_jx/test_edge_1.py
+++ b/tests/test_jx/test_edge_1.py
@@ -558,7 +558,7 @@ class TestEdge1(BaseTestCase):
                 ]
             },
             "expecting_cube": {
-                "meta": {"format": "value"},
+                "meta": {"format": "cube"},
                 "data": {
                     "c": [2, 3, 4, 5, 6, 7, 8],
                     "a": ["x", "y"]
```

Comment (rohit-rk) [inline jx_sqlite/aggs_table.py]: Don't know why, but in my local branch test_union_multiple fails. Tried changes as per your suggestions.

```
ERROR: {
    "data": {
        "a": ["x", "y"],
        "c": ["x", "y"]
    },
    "meta": {"format": "cube"},
    "select": [{"name": "c"}, {"name": "a"}]
} does not match expected {
    "data": {
        "a": ["x", "y"],
        "c": [2, 3, 4, 5, 6, 7, 8]
    },
    "meta": {"format": "cube"}
}
```

Also I am confused for why the output of both "a" and "c" is same and wrong.

Comment (klahnakoski) [inline jx_sqlite/aggs_table.py]: ` test_union_multiple` fails on mine too, it is expected

Comment (klahnakoski) [inline jx_sqlite/aggs_table.py]: the `test_union_multiple` is testing that two separate columns are isolated.  First, check that the result coming out of the database is correct.

File: jx_sqlite/aggs_table.py [hunk 5]
```diff
@@ -377,7 +377,16 @@ def _get(row):
...
                     for json_type, sql in details.sql.items():
-                        concat_sql.append("GROUP_CONCAT(QUOTE(DISTINCT(" + sql + ")))")
+                        concat_sql.append("JSON_GROUP_ARRAY(DISTINCT(" + sql + "))")
+                        query_tables = tables.nest
+                        if details.nested_path not in query_tables:
+                            p = details.nested_path
+                            alias = nest_to_alias[p]
+                            from_sql += (
+                                "\nLEFT JOIN\n (SELECT " + PARENT + ", " + sql + " \nFROM " +
+                                quote_table(concat_field(base_table, p)) + ") " + alias +
```
Comment (klahnakoski) [inline jx_sqlite/aggs_table.py]: You can not use `alias`; choose a name that represents this select clause.  The `alias` could be in use when performing aggregations over an origin that is not the fact table.

**Pull request merged**